charmhelpers/contrib/openstack/cert_utils.py

```diff
@@ -441,3 +448,79 @@ def get_bundle_for_cn(cn, relation_name=None):
         if cert_bundle:
             break
     return cert_bundle
+
+
+def _x509_normalize_input(input_):
+    if isinstance(input_, six.text_type):
+        return input_.encode()
+    if isinstance(input_, bytes):
+        return input_
+    return bytes()
+
+
+def x509_validate_cert_chain(ssl_cert, ssl_ca=None):
+    cmd = ['openssl', 'verify']
+
+    with tempfile.TemporaryFile() as cert_fd, tempfile.NamedTemporaryFile() as ca_fd:
+        cert_fd.write(ssl_cert); cert_fd.seek(0)  # noqa: E702
+        if ssl_ca:
+            ca_fd.write(ssl_ca); ca_fd.seek(0)  # noqa: E702
+            cmd.extend(['-CAfile', ca_fd.name])
+
+        try:
+            subprocess.check_output(cmd, stdin=cert_fd)
```

Comment:
If I'm not mistaken, this always fails when `ssl_cert` consists of a leaf cert + intermediate CA(s). `openstack verify` only reads the first certificate of the chain in the file and ignores the rest.

If we take this openssl CLI approach, you need to split the content of `ssl_ca` into a leaf + the rest of the chain and do:
`openstack verify -CAfile "<ssl_ca>" -untrusted "<the rest of ssl_cert>" "<the first cert in ssl_cert"`.

`openstack verify -CAfile "<ssl_ca>" -untrusted "<ssl_cert>" "<ssl_cert>"` technically works to be simpler even though it's with a duplicate cert in the chain and it doesn't exactly match what Apache2 is doing.

Reply:
I've actually been aware of intermediate certificates posing a minor issue with verification using `openssl`, as you've outlined, which primarily requires in-place splitting of PEM chains and well-considered test cases to properly exemplify check sanity. I just haven't had a sufficient time block to update the proposal to address this.

If there is a way to fix this, I would consider splitting away the leaf as a proper solution.